packages/wp-source/src/libraries/handlers/postType.ts

```diff
@@ -66,12 +67,21 @@ const postTypeHandler = ({
     // 1.2 iterate over finalEndpoints array
     let isHandled = false;
     let isMismatched = false;
-    for (const endpoint of finalEndpoints) {
-      const response = await libraries.source.api.get({
+
+    const promisesAndEndpoints: [
+      string,
+      Promise<Response>
+    ][] = finalEndpoints.map((endpoint) => [
+      endpoint,
+      libraries.source.api.get({
         endpoint,
         params: { slug, _embed: true, ...state.source.params },
-      });
+      }),
+    ]);
 
+    for await (const [endpoint, response] of generateAsyncQueue(
+      promisesAndEndpoints
+    )) {
```

Comment:
Hey @michalczaplinski this looks futuristic 😄  so bear with me and my limited knowledge. Since generators are not part of the browser runtime yet, was looking at how does the `generateAsyncQueue` looks after babel transpiles it(https://babeljs.io/repl#?browsers=and_chr%20%3E%3D%2067%2C%20and_ff%20%3E%3D%2018%2C%20and_uc%20%3E%3D%2011.8%2C%20android%20%3E%3D%2067%2Cnot%20android%20%3C%3D%204.4.4%2Cchrome%20%3E%3D%2049%2Cedge%20%3E%3D%2012%2Cfirefox%20%3E%3D%2018%2Cios_saf%20%3E%3D%2010%2Cnot%20op_mini%20all%2Cop_mob%20%3E%3D%2046%2Copera%20%3E%3D%2036%2Csafari%20%3E%3D%2010%2Csamsung%20%3E%3D%205&build=&builtIns=usage&spec=false&loose=false&code_lz=IYZwngdgxgBAZgV2gFwJYHsICoYHMCmE-ATsMvgILjQCKC-9AFAFAwwAOx6AtqiPiGYBKGAG9WMAPSSYAUQgALYNHwxkC1emKpcqCMAA2MYMVJgY6ODAAKXXvxAwAZDEIATduj3IQAOgnSMAAqCMQQMKjIERDI6MbhJmYWVsgI7AYCADQwAO4axKrqqnCoxCBR-BnchFGWrsBQCmppGRGCbIFFxqbA5npu-AAevsEaMPxQmG6uVTUwBSDoBgBuAmpxXanpmlYA2v1D2Vo6eoYASgIIBsgAuhKTEOUcdnwCFD3mALzPPK9-3MB2IxGOxsqgRJ8AHwwfbZdi7ACMN18RQgwIWEOhsI4uwADDdsgsbkJiQBuZgBGQAYQKZFUwBgACJPEtGckbC8HP42A8ngD2DBvkQcjAALKAkGct4fITkykwM4NekGIx1LqcX4OeLTUDjdCYNRjEplWpEeYCJarEDZMCoSrTIryhZXZCE_DcdCrQ2FFqqHKRJpdSamASeCBuPS4H72Yp2b3dJLKNxOpXGFXsroFAF6SPRv4wJBoIxdZn6gxsviubjsZBgbm5BSoVqMfm-ECoABe-BE4jYPMwT12AGt8GBsu5PN43SAXTdBcYcsBIhzNfhfKQoPgW4DfMtDPQQIwhLKJGxbfaYROvDFp7PyX2YK2BhlyIwR2AT2wAL7ML9AA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=true&timeTravel=false&sourceType=module&lineWrap=true&presets=env&prettier=false&targets=&version=7.12.14&externalPlugins=)

Got scared of the output, lol!

Anyways, my suggestion would be to stick to an `await Promise.all([])`. Would that be suitable for this case? For example something like this looks like it could be fitting:

```js
const results = await Promise.all(
  // Return a map of promises based on each endpoint
  finalEndpoints.map((endpoint) =>
    // The endpoint promise
    libraries.source.api.get({
      endpoint,
      params: { slug, _embed: true, ...state.source.params },
    }).then(res => (
      // Returning the tuple we need
      [endpoint, res]
     ))
  )
);

// Then we can walk over the results to populate
for (const [endpoint, response] of results) {
// ...
```

Again, I might not have the whole context here -- even though I went through the thread and the observablehq link -- but it looks like the `Promise.all` could be enough.

Reply:
https://github.com/frontity/frontity/pull/758#issuecomment-806196882

Continuing here, to keep the discussion on context.

I mean 😄  if we include the iteration part as well the sizes are a bit more night and day:
* [generateAsyncQueue](https://babeljs.io/repl#?browsers=and_chr%20%3E%3D%2067%2C%20and_ff%20%3E%3D%2018%2C%20and_uc%20%3E%3D%2011.8%2C%20android%20%3E%3D%2067%2Cnot%20android%20%3C%3D%204.4.4%2Cchrome%20%3E%3D%2049%2Cedge%20%3E%3D%2012%2Cfirefox%20%3E%3D%2018%2Cios_saf%20%3E%3D%2010%2Cnot%20op_mini%20all%2Cop_mob%20%3E%3D%2046%2Copera%20%3E%3D%2036%2Csafari%20%3E%3D%2010%2Csamsung%20%3E%3D%205&build=&builtIns=usage&spec=false&loose=false&code_lz=IYZwngdgxgBAZgV2gFwJYHsICoYHMCmE-ATsMvgILjQCKC-9AFAFAwwAOx6AtqiPiGYBKGAG9WMAPSSYAUQgALYNHwxkC1emKpcqCMAA2MYMVJgY6ODAAKXXvxAwAZDEIATduj3IQAOgnSMAAqCMQQMKjIERDI6MbhJmYWVsgI7AYCADQwAO4axKrqqnCoxCBR-BnchFGWrsBQCmppGRGCbIFFxqbA5npu-AAevsEaMPxQmG6uVTUwBSDoBgBuAmpxXanpmlYA2v1D2Vo6eoYASgIIBsgAuhKTEOUcdnwCFD3mALzPPK9-3MB2IxGOxsqgRJ8AHwwfbZdi7ACMN18RQgwIWEOhsI4uwADDdsgsbkJiQBuZgBGQAYQKZFUwBgACJPEtGckbC8HP42A8ngD2DBvkQcjAALKAkGct4fITkykwM4NekGIx1LqcX4OeLTUDjdCYNRjEplWpEeYCJarEDZMCoSrTIryhZXZCE_DcdCrQ2FFqqHKRJpdSamASeCBuPS4H72Yp2b3dJLKNxOpXGFXsroFAF6SPRv4wJBoIxdZn6gxsviubjsZBgbm5BSoVqMfm-ECoABe-BE4jYPMwT12AGt8GBsu5PN43SAXTdBcYcsBIhzNfhfKQoPgW4DfMtDPQQIwhLKJGxbfaYROvDFp7PyX2YK2BhlyIwR2AT2wAL7MH_MXlRBqMYgBQ4byB414-POJT6AY4GTjE_wSowV7eJiMISKhN4SAYqAAEakNoAhtugoSbr4gKoL4BDIIwvZ9lhrqnhwJjANwIAAFxiOMBgILg2QAPrunh-BuFxyDEPQ2S-DJ5R0iRZFruwrHsTAX6ZBIX5CBpZIUowoCQLAR6CtC9EAJBwFoC5LlEjAAZe4YIa65ogGG_BznUBBEKQ5BUIZdAMFuQF_KBbjwZBIDHmIP5aaSQA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=true&timeTravel=false&sourceType=module&lineWrap=false&presets=env&prettier=false&targets=&version=7.12.14&externalPlugins=) 10.3 kB
* [Promise.all](https://babeljs.io/repl#?browsers=and_chr%20%3E%3D%2067%2C%20and_ff%20%3E%3D%2018%2C%20and_uc%20%3E%3D%2011.8%2C%20android%20%3E%3D%2067%2Cnot%20android%20%3C%3D%204.4.4%2Cchrome%20%3E%3D%2049%2Cedge%20%3E%3D%2012%2Cfirefox%20%3E%3D%2018%2Cios_saf%20%3E%3D%2010%2Cnot%20op_mini%20all%2Cop_mob%20%3E%3D%2046%2Copera%20%3E%3D%2036%2Csafari%20%3E%3D%2010%2Csamsung%20%3E%3D%205&build=&builtIns=usage&spec=false&loose=false&code_lz=BQQwzgngdgxgBMAlHAvAPjgbwFBzjAeyjABc4AnAUzAFcAbEsVOEAdxAEsyAFcggWw5hKAOhB06wXHjgB6WXABKlEjXJQWcfiAAOcAgDM4OvoOFMARuEoATfRsogYACziUoNnQQ5QS0vAY-4gCiHl4-jCLaOsDA7p7evsjo_jLycAAqzpRuYYlkJgJClKl4dBwW5CDkHNQiYARqMKK6HCIA5irAODK9uQkRADSlMjrVIPxgAFxYcGB0NO2DcAD6lPwWtjMk5DSUyyKHpCAkog1NomNVk3AAvsN9d4giJNlQwFRM6AgjeOnKqnUPnacFeOVUOjoOVYOSglFsvzgAG14uFfMtPgBdX6IRCpPF4RAAbmk0gMBHICEIxDIKLyQwo1C8xEomP0Rk-9EYyEwt2wt0QSCAA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=true&timeTravel=false&sourceType=module&lineWrap=true&presets=env&prettier=false&targets=&version=7.12.14&externalPlugins=) 5.9 kB

But, more related to the async queue processing, thank you so much for explaining it.

My follow-up question is: is there a benefit for the end result to use an async queue? More exactly would the execution go beyond the `for async` loop? I don't think so, but I might be wrong. If that's the case, is there a practical reason to asynchronous populate the `.data`?

Reply:
> I mean 😄 if we include the iteration part as well the sizes are a bit more night and day:
>
> generateAsyncQueue 10.3 kB
> Promise.all 5.9 kB

You're right about that, I didn't consider that! 🙈 

> My follow-up question is: is there a benefit for the end result to use an async queue? More exactly would the execution go beyond the for async loop? I don't think so, but I might be wrong. If that's the case, is there a practical reason to asynchronous populate the .data?

I'm not sure I fully understand the question, in particular I don't know what you mean with "execution go beyond the for async loop?". But let me try to answer what I *think* you mean.

Using the async queue we load data from each of the 3 endpoints in parallel: `post`, `page`, `media`. However, if we were to use `Promise.all`, the data would also load in parallel. So what is the difference? 

If we use the async queue, then we don't have to wait for all 3 requests to resolve before continuing to execute the `postType` handler. So, it's not that there is a benefit to populating `.data` asynchronously but rather that there **is** a benefit to being able to break out of the loop early [here](https://github.com/frontity/frontity/blob/0a1bab4ee07de1eb680bc06d1734d00ed15afc25/packages/wp-source/src/libraries/handlers/postType.ts#L91-L95) **as soon as one of the promises resolves** and this `populated.length > 0` condition is true.